File: solutions/139-word-break/solution.go
package main

import "fmt"

func wordBreak(s string, wordDict []string) bool {
	dp := make([]bool, len(s)+1)
	dp[0] = true
	for i := 1; i <= len(s); i++ {
		for _, word := range wordDict {
			if i >= len(word) && s[i-len(word):i] == word {
				dp[i] = dp[i] || dp[i-len(word)]
			}
		}
	}
	return dp[len(s)]
}

func main() {
	got := wordBreak("leetcode", []string{"leet", "code"})
	fmt.Printf("PASS | %v\n", got)
}

File: solutions/139-word-break/solution.py
from typing import List


class Solution:
    def word_break(self, s: str, word_dict: List[str]) -> bool:
        dp = [False] * (len(s) + 1)
        dp[0] = True
        for i in range(1, len(s) + 1):
            for word in word_dict:
                if i >= len(word) and s[i - len(word) : i] == word:
                    dp[i] = dp[i] or dp[i - len(word)]
        return dp[len(s)]


if __name__ == "__main__":
    sol = Solution()
    got = sol.word_break("leetcode", ["leet", "code"])
    status = "PASS" if got is True else "FAIL"
    print(f"{status} | => {got}")
